Move ElementPropertyReadFile from tensor mechanics to framework, add a function to use it

GiudGiud opened the pull request:
Adds the function to use this UO for reading unstructured mesh data from CSV

Eventually, we may want also an IC and a MaterialProperty
since using a function incurs an extra search for the element (functions work with positions)

GiudGiud committed:
Move read element property from CSV from tensor mechanics to the framework
Refs #8151 #19109

GiudGiud committed:
Add a function to retrieve values from the CSV UO, adapt CSV property reader UO to framework, refs #19109

GiudGiud committed:
Add test for UM CSV function, refs #19109

GiudGiud commented:
Early PR to see what I break by moving and changing the UO, still have to debug something funky with the function

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/850793/) on ad80311 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/19118/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/19118/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 72060375990a6b4756f4964242078db7d1371c82`

GiudGiud committed:
Add documentation, fixup tests a bit, refs #19109

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/850860/) on ece4a8e wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/19118/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/19118/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 72060375990a6b4756f4964242078db7d1371c82`

GiudGiud committed:
Use pointer to UO because UOs are created after functions, so cant use a reference in function constructor
Clang format
refs #19109

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/850905/) on 1da217f wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/19118/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/19118/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 72060375990a6b4756f4964242078db7d1371c82`

GiudGiud commented:
I remember git mv'ing the ElementProperty,..UO files. why is it showing all green

GiudGiud committed:
Address test failures:
allow warnings for older test using CSV files that contain more data than is used
dont rely on default function overrides
fixup link in documentation
clean up function documentation
refs #19109

GiudGiud committed:
Move short time derivative to header for constant function
refs #19109

GiudGiud committed:
Fix up some test failures
TODO: decide on element id when more than 1 is possible for a point
refs #19109

GiudGiud committed:
Use lowest ID when multiple candidates for an element at a point
refs #19109

GiudGiud committed:
Add Node IDs as an option for sorted data in the CSV. Let's keep it continuous for now though
refs #19109

GiudGiud committed:
Adapt tensor mechanics to the renaming of PropertyReadFile
Clang format
Mark initialSetup as override in PieceWiseConstantFromCSV
refs #19109

GiudGiud committed:
Add phantom header to pass CI with old name of (Element)PropertyReadFile
Dont recover ICs tests
Rename documentation
refs #19109

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/860271/) on b0445e5 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>a809cf</th>
      <th colspan=4>#19118 b0445e</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>81.83%</td>
      <td>81.83%</td>
      <td>-0.01%</td>
      <td>77.83%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>71505</td>
      <td>71671</td>
      <td>+166</td>
      <td>165</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15873</td>
      <td>15919</td>
      <td>+46</td>
      <td>47</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/19118/coverage_diff/framework.html)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/19118/coverage/framework)


## Modules coverage

### Tensor mechanics

<table>
  <thead>
    <tr>
      <th></th>
      <th>a809cf</th>
      <th colspan=4>#19118 b0445e</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>87.63%</td>
      <td>87.62%</td>
      <td>-0.01%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>24263</td>
      <td>24161</td>
      <td>-102</td>
      <td>1</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>3424</td>
      <td>3413</td>
      <td>-11</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/19118/coverage_diff/tensor_mechanics.html)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/19118/coverage/tensor_mechanics)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/19118/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/19118/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/19118/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/19118/coverage/contact)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/19118/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/19118/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/19118/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/19118/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/19118/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/19118/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/19118/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/19118/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/19118/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/19118/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/19118/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/19118/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/19118/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/19118/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/19118/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/19118/coverage/richards)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/19118/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/19118/coverage/tensor_mechanics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/19118/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 77.83% is less than the suggested 90.0%



This comment will be updated on new commits.

dschwen commented:
> framework new line coverage rate 77.83% is less than the suggested 90.0%

🧐

dschwen commented on framework/doc/content/source/functions/PiecewiseConstantFromCSV.md:
```diff
@@ -0,0 +1,42 @@
+- defines an interpolation grid, in which case the function will locate the closest point in that interpolation grid, then return the value for that point in the CSV file
+
+
+For the latter case, the first columns of the CSV data must define the coordinates of each point forming the interpolation grid. The number of columns used to define these coordinates
```
is it a grid though? if you're just finding the closest points it might as well be an unstructured point cloud

GiudGiud replied:
changed. Didn't realize grid meant alignment, somehow

dschwen commented on framework/doc/content/source/functions/PiecewiseConstantFromCSV.md:
```diff
@@ -0,0 +1,42 @@
+
+It can assume CSV data
+
+- is sorted by element-id, in which case, when the function is evaluated at a point, it will locate the element containing it then return the value for that element in the CSV file
```
should the first column be the element ID/ node ID? these ID can be non contiguous

dschwen replied:
to clarify, the active elements may have non contiguous IDs

GiudGiud replied:
Ok I added the capability to deal with non-contiguous ID since I think the restriction was making this whole object rather limited in capability.
This took a little more development though, which you will find in the latest commit.

GiudGiud replied:
Removed that commit. Just clear docs

dschwen commented on framework/include/functions/PiecewiseConstantFromCSV.h:
```diff
@@ -0,0 +1,55 @@
+  /**
+   * Get the value of the function (based on time only)
+   * \param t The time
+   * \param pt The point in space (x,y,z) (unused)
```
why is the point in space "(unused)"?  Don't you need that for option 3 above?

GiudGiud replied:
oh yeah time is unused

dschwen commented on framework/include/userobject/PropertyReadFile.h:
```diff
@@ -0,0 +1,142 @@
+   * @param elem the element to get data for
+   * @param prop_num the column index of the property we want to retrieve
+   */
+  Real getElementData(const Elem * elem, unsigned int prop_num) const;
```
what's the difference between this and the method above?

dschwen replied:
I think this API is a bit awkward, but that's just what the legacy code was. Why set a property for an entire element when we are in Voronoi mode? The function should use the point coordinates rather than the element centroids to query values.

GiudGiud replied:
the method above is more general, handles more cases. This one can be called directly if you know (and have checked hopefully) how the data is sorted in your object already. If you are calling this from a random object that doesnt know how the data is in the CSV, then use the one above

dschwen commented on framework/include/userobject/PropertyReadFile.h:
```diff
@@ -0,0 +1,142 @@
+  /**
+   * This function assign properties to element read from file with nearest neighbor / grain based
+   * properties Voronoi centers distribution in the RVE can be Periodic or non-periodic (default)
+   * @param centroid the centroid of the element to get data for
```
why does this have to be a "centrod"? Isn't it just the _point_ you want to evaluate on the voronoi graph?

GiudGiud replied:
true! I ll change that

dschwen commented on framework/include/userobject/PropertyReadFile.h:
```diff
@@ -0,0 +1,142 @@
+  /// Type of read - element, grain, or block (TODO: Make const once apps are patched)
+  ReadTypeEnum _read_type;
+  /// Whether to use a random tesselation for the Voronoi/grain type
+  const bool _use_random_tesselation;
```
? 

This class should focus on reading data from a file. Why is there functionality to perform some random tessellation?!

GiudGiud replied:
because that s what it used to do in Tensor Mecha

It s doing a random tessellation and reading data, assigning it to that random tessellation. It's a general functionality. Probably a rare one outside of TM

GiudGiud replied:
The "periodic grain option" could also be shed. But it doesnt hurt anything, and "periodicity" or "randomness" are not concepts specific to materials stuff

dschwen commented on framework/src/functions/PiecewiseConstantFromCSV.C:
```diff
@@ -0,0 +1,115 @@
+    if (num_candidates == 0)
+      mooseError("No element located at ", p, " to search in element or block sorted CSV values");
+
+    return _read_prop_user_object->getData(min_id_elem, _column_number);
```
where is this handling the block? inside getData?

GiudGiud replied:
yup. The read_type in the function and the UO are the same

GiudGiud commented:
Low-ish coverage rate is mostly:
- error statement that would result from improper use of the APIs (so by developers). This is not a good use of time to try to hit all of them imo
- stuff that is covered in the module (the periodic BC, the grain getValue call directly from the UO, which the function I made bypasses. This does not show up in the coverage report for the framework

I ll try to improve the situation but definitely will not try 100%

loganharbour commented on framework/src/userobject/PropertyReadFile.C:
```diff
@@ -0,0 +1,292 @@
+
+// TODO: this should probably use the built-in min periodic distance!
+Real
+PropertyReadFile::minPeriodicDistance(Point c, Point p) const
```
```suggestion
PropertyReadFile::minPeriodicDistance(const Point & c, const Point & p) const
```

GiudGiud committed:
Address Daniel's review
- clean up doxygen
- have more coverage within framework
- group grain/nearest neighbor stuff
refs #19109

GiudGiud committed:
Address test failure, keep deprecated option for grizzly to pass CI
and rename UO doc file in link to it from function
refs #19109

GiudGiud committed:
Add compatibility option for old param in Grizzly for property reader UO
refs #19109

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/888777/) on 9d04279 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/19118/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/19118/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 03578a069a03e0117c48ba7849b09e4a5ad97c7a`

GiudGiud commented:
@dschwen I should have addressed your review.
The additional capability of being able to index into the CSV file with elem/node/grain/block ID is imo pretty great, but also there is no customer for it right now so it's not optimized.

GiudGiud commented:
@dschwen if you have time this week. 
Last commit is indexing by element ID. If you dont like it I m open to slashing it

loganharbour commented on framework/src/userobject/PropertyReadFile.C:
```diff
@@ -0,0 +1,329 @@
+      data = getVoronoiData(elem->centroid(), prop_num);
+      break;
+
+    // TODO: Delete after Grizzly update
```
Reference an issue num (`idaholab/grizzly#0`)

GiudGiud replied:
https://hpcgitlab.inl.gov/idaholab/grizzly/-/issues/182 so I dont forget

loganharbour commented on framework/src/functions/PiecewiseConstantFromCSV.C:
```diff
@@ -0,0 +1,115 @@
+    (*_point_locator)(p, candidate_elements);
+
+    // Find the element with the lowest ID (that is also on the domain ???????)
+    unsigned int min_elem_id = std::numeric_limits<unsigned int>::max();
+    const Elem * min_id_elem;
+    unsigned int num_candidates = 0;
+
+    std::set<const Elem *>::iterator it = candidate_elements.begin();
+    while (it != candidate_elements.end())
+    {
+      if ((*it)->id() < min_elem_id)
+      {
+        min_elem_id = (*it)->id();
+        min_id_elem = (*it);
+        num_candidates++;
+      }
+      it++;
+    }
+
+    if (num_candidates == 0)
+      mooseError("No element located at ", p, " to search in element or block sorted CSV values");
```
```suggestion
    const Elem * min_id_elem = nullptr;
    for (const auto & elem : candidate_elements)
      if (!elem || elem->id() < min_id_elem->id())
        min_id_elem = elem;
    if (!min_id_elem)
      mooseError("No element located at ", p, " to search in element or block sorted CSV values");
```

This isn't great but I see why you do it

loganharbour commented on framework/src/functions/PiecewiseConstantFromCSV.C:
```diff
@@ -0,0 +1,115 @@
+    std::set<const Elem *> candidate_elements;
+    (*_point_locator)(p, candidate_elements);
+
+    // Find the element with the lowest ID (that is also on the domain ???????)
```
> (that is also on the domain ???????)

This can probably be removed because the check below checks for it

loganharbour commented on framework/src/functions/PiecewiseConstantFromCSV.C:
```diff
@@ -0,0 +1,115 @@
+  else if (_read_type == ReadTypeEnum::VORONOI)
+    // No need to search for the element if we're just looking at nearest neighbors
+    return _read_prop_user_object->getVoronoiData(p, _column_number);
+  else
+  {
+    mooseError("This should not be reachable. Implementation error somewhere");
+    return 0;
+  }
```
```suggestion
  mooseError("This should not be reachable. Implementation error somewhere");
```

mooseError is `[[noreturn]]` so you just erroring is fine

loganharbour reviewed (dismissed):
I'm not sure what is old/new here so take it with a grain of salt, but here's my first run though.

Looks like coverage probably should be increased too (it's re-running now)

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/888836/) on 8694d0c wanted to post the following:


The following coverage requirement(s) failed:
- `contact` coverage rate 78.22% is less than the required 87.0%
- `ray_tracing` coverage rate 92.71% is less than the required 94.0%

GiudGiud commented on framework/include/functions/PiecewiseConstantFromCSV.h:
```diff
@@ -0,0 +1,55 @@
+   */
+  virtual Real value(Real t, const Point & pt) const override;
+
+  /**
+   * Get the time derivative of the function (based on time only)
+   * \param t The time (unused)
+   * \param pt The point in space (x,y,z) (unused)
+   * \return The time derivative of the function at the specified time
+   */
+  virtual Real timeDerivative(Real, const Point &) const override { return 0; };
```
It's valid to take the function time derivative, it's just 0.

loganharbour replied:
IMO you either have a docstring override and say that it returns 0 or remove the docstring completely because the method is overridden and the one from the Function class is sufficient, because `t` and `pt` are used, it's just always zero

GiudGiud replied:
I could say it 'returns 0' in the docstring.
The one from the Function class should be gone. I think that was a bad idea to default to 0 so I removed it

loganharbour replied:
Just leaving this:
```
virtual Real timeDerivative(Real, const Point &) const override { return 0; };
```

is sufficient.

This docstring from `Function.h` will be used:

```
  /**
   * Get the time derivative of the function
   * \param t The time
   * \param p The point in space (x,y,z)
   * \return The time derivative of the function at the specified time and location
   */
```

which is still valid with the whole piecewise constant thing

GiudGiud replied:
oh ok. Yeah I agree with that!

GiudGiud commented on framework/include/userobject/ElementPropertyReadFile.h:
```diff
@@ -0,0 +1,16 @@
+// the old name for PropertyReadFile
+
+#include "PropertyReadFile.h"
+typedef PropertyReadFile ElementPropertyReadFile;
```
this is not for MOOSE, this is for grizzly too
i ll add a note

GiudGiud replied:
or does that do the typedef too ? I ll check

GiudGiud replied:
https://civet.inl.gov/job/908129/ it doesnt build without the typedef. I have a registerObjectRenamed already

GiudGiud commented:
coverage failure is unrelated to the PR btw. This does not touch ray tracing or contact

loganharbour commented:
It is related, it’s just an old run before RT and contact were improved. Framework diff is valid

cticenhour reviewed (dismissed):
Please add a newsletter entry for December, to broadcast the deprecation of `ElementPropertyReadFile`. I am merging a template as part of #19520

GiudGiud commented on framework/doc/content/source/functions/PiecewiseConstantFromCSV.md:
```diff
@@ -0,0 +1,45 @@
+
+!alert note
+When use data by block or by element, if there is multiple possibilities for the element to choose from, for example at a node,
+the element with the lowest ID will be used.
```
Not really. Imo I kind of expect a lot of indetermination in that kind of thing when the data doesnt match one to one.  Even the nearest neighbor stuff could have indetermination

GiudGiud replied:
Ok I just added a warning

GiudGiud committed:
Address Logan s review:
- more const
- use bounding box over keeping each attribute separate
- use a namespace for an enum
refs #19109

GiudGiud commented on framework/include/userobject/PropertyReadFile.h:
```diff
@@ -0,0 +1,145 @@
+  Point _top_right;
+  Point _bottom_left;
+  Point _range;
+  Real _max_range;
```
changed

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/908079/) on 7b910ed wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/19118/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/19118/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 4c02736557d6f1a333ab026f505973bf6eb5bc6c`

GiudGiud committed:
Clang format
Add newsletter post for new capabilities
Need the typedef for Grizzly
Fix the short loop for min element id
refs #19109

GiudGiud commented:
Should be good for another round if it passes tests

loganharbour commented on framework/include/userobject/PropertyReadFile.h:
```diff
@@ -0,0 +1,145 @@
+/**
+ * How data is organized in the CSV file
+ */
+enum class ReadTypeEnum
+{
+  ELEMENT = 0,
+  VORONOI = 1,
+  BLOCK = 2,
+  NODE = 3,
+  GRAIN = 4
+};
```
uh why not put it in the class...?

GiudGiud replied:
I need it for the function too. If I put it in the class I m going to have to write an accessor etc. 
I mean it s totally fine if you prefer that. I ll address it with your next set of reviews

GiudGiud replied:
well actually at param creation time I wont have the other class initialized. So I d have to inherit from it if I wanted access to it?

dschwen replied:
No, its a type, you'd have access through `PropertyReadFile::ReadTypeEnum`

GiudGiud replied:
Ok I did it this way instead of having its own namespace.

GiudGiud commented:
yup ready for another round

GiudGiud committed:
Make some range checking attributes private, keep the one needed by BlackBear protected, refs #19109

GiudGiud committed:
Add more useful errors for non-consecutive issues, refs #19109

GiudGiud committed:
Add better documentation to PropertyReadFile, better examples & clear warning on consecutiveness, refs #19109

GiudGiud committed:
Add warning in PiecewiseConstantFromCSV when multiple elements are found when using elements for the CSV file, refs #19109

GiudGiud committed:
Move read_type  enum from a namespace to living inside PropertyReadFile class, refs #19109

GiudGiud commented:
@dschwen I removed a commit that was addressing the weakness you pointed out (need for consecutive numbers) and instead properly documented the limitation. 

@dschwen @loganharbour down for another review, I'd like to get this in

GiudGiud commented on framework/include/userobject/PropertyReadFile.h:
```diff
@@ -0,0 +1,145 @@
+
+private:
+  unsigned int _nelem;
+  unsigned int _nnodes;
```
That's legacy. This object tracks the number of elements/block/etc, mostly because it doesnt support non-consecutive indexing

GiudGiud committed:
Merge branch 'next' into PR_heist_tm

GiudGiud commented:
I ll remove the merge commit when I get a review

GiudGiud committed:
Handle hacky AD value function API, refs #14262

dschwen reviewed (approved):
Fine with me

loganharbour commented:
This was not ready to merge - please don't dismiss my reviews: https://civet.inl.gov/job/1024376/

I was in the process of resetting testing

cticenhour commented:
@GiudGiud can you move the news entry for this into 2022_04.md?

GiudGiud commented:
will do, I ll catch that in an upcoming maintenance PR
